23 activity to take picture and title it

iliasmerigh opened the pull request:
### General description of work done

In this PR I finished the following things :

- Created an activity to take a picture
- Start the camera to take the picture we want to share
- Once the picture is taken, we can title it
- Added a fake "Share" button that we will implement once the database will be ready
- Tested everything

### Encountered issue

- Espresso allows us to automatically grant the user permissions for testing but it doesn't allow us to deny them. So I didn't manage to test the path were the camera permissions are denied. I tried to add dependency injection but it seems that UI tests don't have access to code so it not possible to inject parameters (it only depends on the environnement).
Look at my discord post for more details : https://discord.com/channels/1076146742615097397/1087462731831066675/1087462731831066675

- Tried to test that an imageView displays the correct image but there is no matcher for that. We should think about creating a custom matcher.

- Tried to test Exceptions but there was a huge bug were Exception thrown during activity launching are not caught. Finally, i removed the Exceptions are finished the activity instead. Moreover, the Exceptions handling in UITest doesn't work properly with the `assertThrows` or the `@Test(expected = ...)`. If you want to catch exceptions in UI tests, you should do as follows : 

```
assertThrows(IllegalStateException.class, () -> {
        try {
               onView(withId(R.id.myCrazyButton)).perform(click());
        } catch (PerformException e) {
               throw e.getCause(); // Exceptions are nested so we have to unnest them
        }
});
```

Look at my discord post for more details : https://discord.com/channels/1076146742615097397/1086885202430787635/1086885202430787635

iliasmerigh commented:
Please look at the method displayPostImage() in SharePostActivity.

We have an activity AddNewPostActivity that start another activity SharePostActivity by giving it the image to post (as an extra in the intent).

If the image is not given correctly, an IllegalArgumentException is raised when the image should be displayed during the activity launching. BUT it seems that thrown exceptions during activity LAUCHING cannot be caught by UI test. Only exceptions thrown on a view action (ex: click) are caught. Thus I decided to do like that :

If the SharePostActivity is started without an image as extra, a fake image automatically is displayed (because we cannot handle an exception during launching). The exception is thrown only when the user clicks on the "Share button".

Otherwise we can throw an IllegalArgmentException if the image cannot be displayed (nor shared) but it will never be tested.

Please tell me what do you think about that. If you find how to handle exceptions during launching from the UI tests please let me know, but I have spent more than 10 hours on it and I didn't find anything. I've spent so much time because we will probably face the same problem in the future so we cannot ignore it.

2Tricky4u reviewed (changes requested):
Overall good PR, still some details to fix and maybe we should convert the activity as a fragment as another task. Good job!

2Tricky4u commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/SharePostActivity.kt:
```diff
@@ -0,0 +1,106 @@
+
+        var image = getPostImage()
+
+        if (image != null) {
+            postImageView.setImageBitmap(image)
+        } else {
+            finish()
+        }
```
```suggestion
       image?.let { postImageView.setImageBitmap(it) } ?: finish()
```

iliasmerigh replied:
Actually, I prefer your version when we want to assign a value to a variable, but when we have two operations (void) I prefer to keep the old version which is more understandable and clear.

2Tricky4u commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/SharePostActivity.kt:
```diff
@@ -0,0 +1,106 @@
+        // If the intent that launched the activity doesn't have the correct extra, "image" will be null.
+        // If so, we finish the activity. Otherwise we can display the image.
+
+        var image = getPostImage()
```
```suggestion
        val image = getPostImage()
```

iliasmerigh replied:
Well seen !

iliasmerigh commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/SharePostActivity.kt:
```diff
@@ -0,0 +1,106 @@
+        // If the intent that launched the activity doesn't have the correct extra, "image" will be null.
+        // If so, we finish the activity. Otherwise we can display the image.
+
+        var image = getPostImage()
```
Indeed !

iliasmerigh commented on app/src/test/java/com/github/sdp_begreen/begreen/entities/PostTest.kt:
```diff
@@ -0,0 +1,23 @@
+    private val post = Post(testTitle, mockBitmap)
+
+    @Test
+    fun testGetTitle() {
```
Yes its better to specify the expected behavior. I made the change.

iliasmerigh commented on app/src/androidTest/java/com/github/sdp_begreen/begreen/activities/SharePostActivityTest.kt:
```diff
@@ -0,0 +1,119 @@
+            .perform(click())
+
+        // Check that the activity has finished, so we go back to the previous
+        assertTrue(activity.state.isAtLeast(Lifecycle.State.DESTROYED))
+
+        activity.close()
```
You are right, I will remove the activity.close()

iliasmerigh commented on app/src/androidTest/java/com/github/sdp_begreen/begreen/activities/AddNewPostActivityTest.kt:
```diff
@@ -0,0 +1,174 @@
+package com.github.sdp_begreen.begreen.activities
```
I mixed them because the test names were too long. But you are right, I've found better tests names and we should stay on camelCase

iliasmerigh commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/SharePostActivity.kt:
```diff
@@ -0,0 +1,106 @@
+
+        var image = getPostImage()
+
+        if (image != null) {
+            postImageView.setImageBitmap(image)
+        } else {
+            finish()
+        }
```
Actually, I prefer your version when we want to assign a value to a variable, but when we have two operations (void) I prefer to keep the old version which is more understandable and clear.

iliasmerigh commented on app/src/androidTest/java/com/github/sdp_begreen/begreen/activities/AddNewPostActivityTest.kt:
```diff
@@ -0,0 +1,174 @@
+        // When debbuging, we noticed that for this test the value of state was RESUMED, while the state
+        // of the test onActivityResult_OK_FromCameraDisplaysShareActivity was CREATED. The state seems to be independent
+        // of the time since we tested at different instants with thread.sleep.
+        assertTrue(activityRule.scenario.state.isAtLeast(Lifecycle.State.RESUMED))
```
You are right. I've applied the changes. Also, I've added robustness to the test by checking for strict equality instead of inequality.

iliasmerigh commented on app/src/androidTest/java/com/github/sdp_begreen/begreen/activities/SharePostActivityTest.kt:
```diff
@@ -0,0 +1,119 @@
+        Intents.release()
+    }
+
+    @Test
+    fun intentWithNoExtraImageThrowsIllegalArgumentException() {
+
+        val intent = Intent(ApplicationProvider.getApplicationContext(), SharePostActivity::class.java).apply {
+            this.putExtra("Hello", "Hello")
+        }
+        val activity = launchActivity<SharePostActivity>(intent)
+
+        assertTrue(activity.state.isAtLeast(Lifecycle.State.DESTROYED))
+    }
+
+    fun intentWithExtraNotBitmapThrowsIllegalArgumentException() {
+
+        val intent = Intent(ApplicationProvider.getApplicationContext(), SharePostActivity::class.java).apply {
+            this.putExtra(AddNewPostActivity.EXTRA_IMAGE_BITMAP, "Hello")
+        }
+        val activity = launchActivity<SharePostActivity>(intent)
+
+        assertTrue(activity.state.isAtLeast(Lifecycle.State.DESTROYED))
+    }
+
+    fun intentWithNullExtraThrowsIllegalArgumentException() {
+
+        val intent = Intent(ApplicationProvider.getApplicationContext(), SharePostActivity::class.java).apply {
+            this.replaceExtras(null)
+        }
+        val activity = launchActivity<SharePostActivity>(intent)
+
+        assertTrue(activity.state.isAtLeast(Lifecycle.State.DESTROYED))
+    }
+
```
1) At first I was throwing an Exception but I changed the behavior of the class and forgot to change the test name. I made the change by replacing "ThrowsIllegalArgumentException" by "FinishesActivity".

2) I will switch to Hamcrest

3) Well seen. I've changed to a strict equality : assertThat(activity.state, equalTo(Lifecycle.State.DESTROYED))

iliasmerigh commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/SharePostActivity.kt:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * Helper function to get the whole post instance
+     */
+    private fun getPost() : Post? {
```
Actually, the project contains an incoherence regarding this class : I am working on the activity to take the photo of the post and title it, while someone else is working on the get/set of a photo in the database. I will keep this name for now and I will change it once I will merge my work with the database. In other words, there is a probability that this class will be deleted or merged with another one once the database will be ready.

iliasmerigh commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/SharePostActivity.kt:
```diff
@@ -0,0 +1,106 @@
+     * Helper function to return the image post which is an extra of the intent
+     */
+    private fun getPostImage() : Bitmap? {
+
+        val image = intent.extras?.get(AddNewPostActivity.EXTRA_IMAGE_BITMAP) as? Bitmap
+
+        return image
```
Well seen. I've changed it in order to handle the backward compatibility. However, it decreases the test coverage because a path cannot be tested since it depends on the environnement (the SDK version). But it's not a big deal.

iliasmerigh commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/AddNewPostActivity.kt:
```diff
@@ -0,0 +1,93 @@
+            // Start the camera intent. Actually, it handles the permission checks by its own so the previous
+            // verification is not mandatory. But keep it if android decides to change the behavior of the camera permission.
+            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
+            startActivityForResult(intent, REQUEST_IMAGE_CAPTURE)
```
Thank you for the reference, I've applied the changes. I've also changed the tests accordingly.

iliasmerigh commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/AddNewPostActivity.kt:
```diff
@@ -0,0 +1,93 @@
+
+    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+
+        super.onActivityResult(requestCode, resultCode, data)
```
Yes thank you I've applied the changes

iliasmerigh commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/AddNewPostActivity.kt:
```diff
@@ -0,0 +1,93 @@
+        super.onActivityResult(requestCode, resultCode, data)
+
+        // When we receive the photo from the camera, we start a new activity to share it
+        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == RESULT_OK && data != null) {
+
+            // Get the image from the camera activity
+            val image = data.extras?.get("data") as? Bitmap
+
+            if (image != null) {
+
+                // Start the SharePost activity with the taken image
+                startSharePostActivity(image)
+            }
+        }
```
Yes that's what I have done by implementing a callback function that will be executed once the camera activity will return the result. But unfortunately, the callback function needs to be **non-private** otherwise it cannot be tested. But it's a good improvement because this way of handling activities' result makes the code far more easy and scalable. I will reuse it for sur in the future, thank you for pointing it !

iliasmerigh commented on app/src/main/java/com/github/sdp_begreen/begreen/entities/Post.kt:
```diff
@@ -0,0 +1,10 @@
+package com.github.sdp_begreen.begreen.entities
+
+import android.graphics.Bitmap
+
+class Post(private val title: String, private val image: Bitmap) {
+
+    fun getTitle(): String = title
+
+    fun getImage(): Bitmap = image
+}
```
Yes it's exactly what I've answered in one of your previous comment. Glad we think the same !

iliasmerigh commented on app/src/main/res/layout/activity_add_new_post.xml:
```diff
@@ -0,0 +1,19 @@
+        android:id="@+id/addNewPostBtn"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="Add new post"
```
Thank you I didn't know this trick. I've changed it.

iliasmerigh commented on app/src/test/java/com/github/sdp_begreen/begreen/entities/PostTest.kt:
```diff
@@ -0,0 +1,23 @@
+
+class PostTest {
+
+    private val mockBitmap = mock(Bitmap::class.java)
```
I used a mock to check that the image getter returns the expected bitmap. Should I change it to a fake value ? I though that the purpose of the mock library was automatically prevent this fake instances.  

bitmap = Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888)

iliasmerigh replied:
Anyway, I've changed because I was not confortable with the mock. I don't know was it was underlined in red even if the tests passed. But I am still curious about why you suggested to change it

iliasmerigh commented on app/src/main/res/layout/activity_share_post.xml:
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Thank you for pointing it. It's fixed !

david-schroeter commented on app/src/test/java/com/github/sdp_begreen/begreen/models/PostTest.kt:
```diff
@@ -0,0 +1,24 @@
+
+class PostTest {
+
+    private val testBitmap = mock(Bitmap::class.java)
```
I am not sure that you should mock the `Bitmap` class here.

Mock testing are used to isolate the class under test by injecting fake dependency (mocked dependency) into this class, so that we don't depends on external implementation at test time.

An example would be if one of your class connect to the internet through an `HttpClient` interface, then in unit test you don't want to actually use the real implementation of the `HttpClient` you would rather `mock` the `HttpClient` to return fake response only for test time, to isolate only the logic of the class you are currently testing, not the behaviour of the `HttpClient` implementation.

For a better explanation, maybe you can have a look to the [test lecture](https://github.com/sweng-epfl/public/tree/main/lectures/Testing#how-can-one-test-entire-modules) from last semester that we had in Sweng.

But here, the way you are using it you are not actually mocking the `Bitmap` class, you are indeed creating a mock instance of `Bitmap` but you don't inject ii aywhere and you don't explicitely mock methods of the `Bitmap` class.

Here I think you simply need to create an fake `Bitmap` instance just to store it in the `Post` instance that you are testing.

Let me know if you have any questions about what I said.

iliasmerigh replied:
Thank you for the explanations. I've changed the mock with :

`private val testBitmap = Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888)`

This forced me to move the test to the androidTest folder otherwise the Bitmap is not recognized. I saw that somebody else had the same issue with the _UserTest_ and did the same.

david-schroeter commented on app/build.gradle:
```diff
@@ -67,9 +67,15 @@ dependencies {
     implementation 'com.google.firebase:firebase-auth-ktx'
     implementation 'com.google.firebase:firebase-storage-ktx'
 
+    // Mockito library for testing
+    testImplementation 'org.mockito:mockito-inline:5.2.0'
+    androidTestImplementation 'org.mockito:mockito-core:5.2.0'
+    androidTestImplementation 'org.mockito.kotlin:mockito-kotlin:4.1.0'
+    androidTestImplementation 'org.mockito:mockito-android:5.2.0'
```
If we end up not using mock testing in [PostTest](https://github.com/SDP-BeGreen/BeGreen/blob/af44248df800de0b0205f4ba0763b57673f6ee9c/app/src/test/java/com/github/sdp_begreen/begreen/models/PostTest.kt), I think it would be better to remove those dependency, no to have unused dependency in our gradle. We can always add them later when we need them.

david-schroeter reviewed (changes requested):
First of all nice PR, thanks for the camera implementation.

Then thanks as well for applying some modification that @2Tricky4u recommanded to you, your changes looks good.

I just have on thing to add regarding the way you are using `Mock testing` in you file [PostTest](https://github.com/SDP-BeGreen/BeGreen/blob/af44248df800de0b0205f4ba0763b57673f6ee9c/app/src/test/java/com/github/sdp_begreen/begreen/models/PostTest.kt).

I wrote a comment there, let me know what you think of it.

iliasmerigh commented:
The _AddNewPostActivity_ test coverage dropped a bit because I added some checks in the code which made some paths unreachable (i.e check the SDK version to ensure backward compatibility). Anyway, I am currently working on my second task where I changed _AddNewPostActivity_ to a fragment so this class (with its tests) are going to be deleted (replaced). I will try to increase the coverage again with the fragment version, but almost all non-tested paths depend on the environnement (camera permissions, sdk version, etc...) so the current coverage is almost at its maximum.

david-schroeter reviewed (approved):
Thanks for the changes it looks all good to me.

If its also ok for @2Tricky4u then I thinks it's all good.

2Tricky4u commented:
It looks good to me too

2Tricky4u reviewed (approved):
It looks good to me. Well done!

SolalPirelli commented on app/src/androidTest/java/com/github/sdp_begreen/begreen/models/PostTest.kt:
```diff
@@ -0,0 +1,24 @@
+
+    private val testBitmap = Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888)
+    private val testTitle = "Test Title"
+    private val post = Post(testTitle, testBitmap)
```
What should happen if the title is empty? Or composed only of whitespace?

SolalPirelli commented on app/src/main/java/com/github/sdp_begreen/begreen/models/Post.kt:
```diff
@@ -0,0 +1,10 @@
+
+import android.graphics.Bitmap
+
+class Post(private val title: String, private val image: Bitmap) {
```
Looks like you could omit `private` here and omit the explicit `get...` functions, since `val` is immutable?

SolalPirelli commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/AddNewPostActivity.kt:
```diff
@@ -0,0 +1,114 @@
+        } else {
+
+            // Start the camera intent. Actually, it handles the permission checks by its own so the previous
+            // verification is not mandatory. But keep it if android decides to change the behavior of the camera permission.
```
Is there some reason to believe Android might change in this specific way soon? (Android could change in all kinds of ways, it's impossible to predict them all)

SolalPirelli commented on app/src/main/java/com/github/sdp_begreen/begreen/activities/SharePostActivity.kt:
```diff
@@ -0,0 +1,111 @@
+     */
+    private fun getPostImage() : Bitmap? {
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```
This is the 2nd time that code appears in as many files, perhaps it's worth extracting that logic into a "compatibility" utility somewhere
